core/src/main/kotlin/net/corda/core/crypto/AlgorithmManager.kt

```
@@ -0,0 +1,108 @@
+    )
+
+    /** EdDSA signature scheme using the ed255519 twisted Edwards curve. */
+    private val EDDSA_SIGNATURE = SignatureAlgorithm(
+            EdDSAEngine(),
+            net.i2p.crypto.eddsa.KeyPairGenerator(), // EdDSA engine uses a custom KeyPairGenerator Vs BouncyCastle.
+            EdDSAGenParameterSpec(EdDSANamedCurveTable.CURVE_ED25519_SHA512),
+            256,
+            "EdDSA",
+            "EdDSA signature scheme using the ed255519 twisted Edwards curve"
+
+    )
+
+    /** SPHINCS-256 hash-based signature scheme. It provides 128bit security against post-quantum attackers. */
+    private val SPHINCS_SIGNATURE = SignatureAlgorithm(
+            Signature.getInstance("SHA512WITHSPHINCS256","BCPQC"),
+            KeyPairGenerator.getInstance("SPHINCS256", "BCPQC"),
+            SPHINCS256KeyGenParameterSpec(SPHINCS256KeyGenParameterSpec.SHA512_256),
+            256,
+            "SPHINCS-256",
+            "SPHINCS-256 hash-based signature scheme. It provides 128bit security against post-quantum attackers."
+    )
+
+    /** Supported signature algorithms */
+    private val supportedAlgorithms = hashMapOf(
+            RSA_SIGNATURE.keyAlgorithm to RSA_SIGNATURE,
+            ECDSA_SIGNATURE.keyAlgorithm to ECDSA_SIGNATURE,
+            EDDSA_SIGNATURE.keyAlgorithm to EDDSA_SIGNATURE,
+            SPHINCS_SIGNATURE.keyAlgorithm to SPHINCS_SIGNATURE
+    )
+
+    /**
+     * Factory pattern to retrieve the corresponding [SignatureAlgorithm] based on the type of the [String] input.
+     * This function is usually called by key generators.
+     * In case the input is not a key in the supportedAlgorithms map, null will be returned.
+     * @param algorithmName a [String] that should match a key in supportedAlgorithms map.
+     * @return a SignatureAlgorithm of the ones currently supported or null.
+     */
+    fun findAlgorithm(algorithmName: String): SignatureAlgorithm? {
```

This could be `fun findAlgorithm(name: String): SignatureAlgorithm? = supportedAlgorithms[name]`

I wonder if returning null is the right choice here. Probably we should throw an exception instead? That way this function would add some value over the underlying Map.get method. I imagine that in most cases where this function will be used, the algorithm not being found is a showstopper anyway.